### ISSUE TITLE
Avoid link attributes, use common CRUD pattern

### ISSUE THREAD

**Issue by chrysn** [core-wg/dynlink#16]
As discussed on Saturday, it'd be great if dynlink could manage to do without non-target link attributes. At the same time, manipulating individual links in a link-format document is tricky. The proposed interface is explained most quickly by examples:

Throughout the example, we want to replicate data from coap://a/thermometer to coap://b/temperature-input; it is immaterial to it whether we post the example to a binding table on a, b or a third party (the URIs might just get more compact). The refresh-time is a purely illustrative example of a literal-valued extension property we may want to ascribe to a binding. It assumes we've already found a binding table (henceforth called binding collection) at /binding-collection.

```
POST /binding-collection/
Content-Format: (application/link-format)
Payload:
<coap://a/thermometer>;rel="binding-source",
<coap://b/tmperatuer-input>;rel="binding-target",
<>;rel=self;bind-source=observe;bind-dest=put;refresh-time=30

2.01 Created
Location: /binding-collection/1
```

This allows all attributes to be pure, allows individual bindings to be addressed by URI (and thus, for example, to be deleted one-by-one even before there is a patch format for links).

All of this is unrelated to CoRAL, but can be expressed a bit more smoothly there, so let's have the example repeated in CoRAL (note that if we go all the way, the `rel=binding-source` / `#using iana binding-source` would probably be replaced by `rel=urn:...:dynlink:source` / `#using dynlink source`):

```
POST /binding-collection/
Content-Format: (text/coral)
Payload:
#using iana ...
#prefix dynlink ...
iana:binding-source <coap://a/thermometer>
iana:binding-target <coap://b/tmperatuer-input>
dynlink:source-method dynlink:method:observe
dynlink:dest-method dynlink:method:put
dynlink:refresh-time 30

2.01 Created
Location: /binding-collection/1
```

To complete the example, reading the binding table could happen in a deep or shallow way (don't know what's a good default):

```
GET /binding-collection/

2.05 Content
Content-Format: (text/coral)
Payload:
#using iana ...
#using collections ...
#using application-specific ...
collections:item </binding-collection/1> {
    iana:binding-source <coap://a/thermometer>
    ...
    application-specific:created-by <http://users.example.com/~chrysn/#me>
}
collections:item </binding-collection/2> { ...}
```

The server has, in this example, added additional metadata statements to the item -- also something that would not have previously been possible.

**Comment by chrysn** [comment]
This, by the way, also solves the long-standing comment that link-format can't have links like `</local/42>;anchor="coap://elsewhere/remote/foo"` because that'd be resolved as coap://elsewhere/local/42 -- you have to write `<coap://my-local-hostname/local/42>;anchor=...` for those. This limitation is even there in CoRAL, so the two links are really required unless we want to impose knowledge of the own authority component to the client.

**Comment by bsilverajan** [comment]
The binding table indeed does need some clarification on :

1. Why the resource collection as indicated in the current draft contains weblinks instead of items in the collection
2. What changes are necessary in order to allow item manipulation if we indeed have a resource collection

So this proposal does solve both problems. Having alignment with CoRAL is a plus. This seems possible to tweak also for batch operations.

**Comment by bsilverajan** [comment]
One comment I had was that taking your proposal on-board would require 3 links in the request payload.
Is the rationale to explicitly indicate the relation type to the target as source or destination, a limitation to CoRAL-based expression, or are you proposing it for easy manipulation (such as a possible replace operation in future with iPatch)?

**Comment by chrysn** [comment]
The explicit relation types to source and destination are there because the information model of links is not friendly to sequence. Furthermore, a relation type needs to be present anyway, and is very small (one byte) in CoRAL.

Or was the question more about why the third link is there? That's because we assign (literal-valued) properties to the binding itself, which is done by a self-link in link-format or by just adding them to the document in CoRAL. Those details (like the binding type) are neither a property of the source (for then they would apply to all bindings from that source) nor of the destination (for then they would apply to all bindings *to* that source), but to this particular dynlink between them.

**Comment by bsilverajan** [comment]
> Or was the question more about why the third link is there? That's because we assign (literal-valued) 

No, the question was about the first two links as it was clear to me what the third link represented. However, it became clearer to me when I re-read that your proposal makes it explicit as the entry could be posted to a binding table that could theoretically even reside in a third party server.
